src/main/java/uk/gov/companieshouse/overseasentitiesapi/service/BeneficialOwnerChangeService.java

```diff
@@ -0,0 +1,435 @@
+        .getBeneficialOwnersCorporate();
+    return beneficialOwnersGovernmentOrPublicAuthority
+        .stream()
+        .map(this::covertBeneficialOwnerCorporateChange)
+        .filter(Objects::nonNull)
+        .collect(Collectors.toList());
+  }
+
+  /**
+   * Converts a BeneficialOwnerCorporateDto into a BeneficialOwnerChange object.
+   *
+   * <p>The method detects changes between the provided BeneficialOwnerCorporateDto and its
+   * corresponding information in the public and private data.</p>
+   *
+   * @param beneficialOwnerCorporateDto The BeneficialOwnerCorporateDto object to be converted.
+   * @return A BeneficialOwnerChange object if changes are detected, otherwise null.
+   */
+  public BeneficialOwnerChange<CorporateBeneficialOwnerPsc> covertBeneficialOwnerCorporateChange(
+      BeneficialOwnerCorporateDto beneficialOwnerCorporateDto) {
+
+    var beneficialOwnerChange = new CorporateBeneficialOwnerChange();
+
+    var psc = new CorporateBeneficialOwnerPsc();
+    beneficialOwnerChange.setPsc(psc);
+
+    Pair<PscApi, PrivateBoDataApi> publicPrivateBoPair = publicPrivateBo.get(
+        beneficialOwnerCorporateDto.getChipsReference());
+
+    if (publicPrivateBoPair == null) {
+      return null;
+    }
+
+    if (publicPrivateBoPair.getRight() != null) {
+      beneficialOwnerChange.setAppointmentId(publicPrivateBoPair.getRight().getPscId());
+    }
+
+    ChangeManager<CorporateBeneficialOwnerPsc, PscApi, PrivateBoDataApi> changeManager = new ChangeManager<>(
+        psc, publicPrivateBoPair);
+
+    var beneficialOwnerNatureOfControlTypes = beneficialOwnerCorporateDto.getBeneficialOwnerNatureOfControlTypes();
+    var trusteesNatureOfControlTypes = beneficialOwnerCorporateDto.getTrusteesNatureOfControlTypes();
+    var nonLegalFirmMembersNatureOfControlTypes = beneficialOwnerCorporateDto.getNonLegalFirmMembersNatureOfControlTypes();
+
+    var collectedNatureOfControl = NatureOfControlTypeMapping.collectAllNatureOfControlsIntoSingleList(
+        beneficialOwnerNatureOfControlTypes,
+        trusteesNatureOfControlTypes, nonLegalFirmMembersNatureOfControlTypes);
+
+    boolean hasChange = setCommonAttributes(changeManager,
+        publicPrivateBoPair,
+        beneficialOwnerCorporateDto.getServiceAddress(),
+        beneficialOwnerCorporateDto.getPrincipalAddress(),
+        collectedNatureOfControl);
+
+    hasChange |= changeManager.compareAndBuildLeftChange(
+        beneficialOwnerCorporateDto.getName(),
+        PscApi::getName,
+        CorporateBeneficialOwnerPsc::setCorporateName);
+
+    Identification publicIdentification = publicPrivateBoPair.getLeft() == null ? null
+        : publicPrivateBoPair.getLeft().getIdentification();
+
+    ChangeManager<CompanyIdentification, PscApi, Identification> companyIdentificationChangeManager
+        = new ChangeManager<>(psc.getCompanyIdentification(),
+        new ImmutablePair<>(publicPrivateBoPair.getLeft(), publicIdentification));
```

Comment:
If `publicIdentification` is null (line 189), would we still want to continue trying to retrieve data from the object below?

Reply:
Ok, fair enough. Thanks for clarifying!